cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -77,6 +83,91 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
                     )
 
 
+def _build_gateset_and_gate_durations(
+    proto: v2.device_pb2.DeviceSpecification,
+) -> Tuple[cirq.Gateset, Dict[cirq.GateFamily, cirq.Duration]]:
+    gates_list: List[Union[Type[cirq.Gate], cirq.Gate, cirq.GateFamily]] = []
+    fsim_gates: List[Union[Type[POSSIBLE_FSIM_GATES], POSSIBLE_FSIM_GATES]] = []
+    gate_durations: Dict[cirq.GateFamily, cirq.Duration] = {}
+
+    # TODO(#5050) Describe how to add/remove gates.
+
+    for gate_spec in proto.valid_gates:
+        gate_name = gate_spec.WhichOneof('gate')
+        cirq_gates: List[Union[Type[cirq.Gate], cirq.Gate, cirq.GateFamily]] = []
+
+        if gate_name == 'syc':
+            cirq_gates = [ops.SYC]
+            fsim_gates.append(ops.SYC)
+        elif gate_name == 'sqrt_iswap':
+            cirq_gates = [cirq.SQRT_ISWAP]
```

Comment:
> IIUC including the gate, while redundant, doesn't change gate validation behavior.

The purpose of a gateset is description and validation. Having redundant gates makes description more confusing and verbose, which should be avoided. 

> Different FSimGates have different durations today:

We can insert a separate fsim gate family instance for each fsim gate type in DeviceSpecification; so we can associate gate duration with the corresponding fsim gate family. i.e.

```python
# gate durations dict should contain this.
{cirq.FSimGateFamily(gates_to_accept=[cirq.SQRT_ISWAP]):  32_000,
cirq.FSimGateFamily(gates_to_accept=[cg.SYC]):  12_000}
# instead of .
{cirq.GateFamily(cirq.SQRT_ISWAP):  32_000,
cirq.GateFamily(cg.SYC):  12_000}
```

In general, we should be consistent with the gate families that we are using for description and validation, so that inconsistencies like the following cannot occur:

```python
$> sqrt_iswap_gate = cirq.FSimGate(-np.pi/4, 0)
# Returns True because of validation across types by FSimGateFamily.
$> assert sqrt_iswap_gate in sqrt_iswap_metadata.gateset 
# Should return True but will return False right now because `sqrt_iswap_gate in cirq.GateFamily(cirq.SQRT_ISWAP)` is False
$> assert any(sqrt_iswap_gate in gf for gf in sqrt_iswap_metadata.gate_durations) 
```

Reply:
As discussed offline, this makes a lot of sense, and I'll push for a [change](https://github.com/quantumlib/Cirq/pull/5427) in the `GridDeviceMetadata.gate_duration`. Didn't end up using a function, but the proposal captures the same spirit.

Reply:
Actually your approach currently fails `GridDeviceMetadata` validation because all gate duration keys are expected to be in the gateset. Let's discuss in #5427 since it potentially involves a change in `GridDeviceMetadata`. Are you onboard with keeping the gate duration the way it is and leaving a TODO for followup?